### PR TITLE
Update cmake/README.md for #783 PR

### PR DESCRIPTION
**Pull request by podsvirov**
The documentation for the CMake project, corresponding to the changes presented in the #783 PR.
Documentation may require linguistic corrections.
If #783 PR is merged, this documentation should also be merged.

### CODE REVIEW

File: cmake/README.md [hunk 1]
```diff
@@ -1,60 +1,243 @@
...
-   PATH).
-10. Copy libprotobuf.lib, libprotobuf-lite.lib, and libprotoc.lib wherever you
-    put libraries.
-
-  To avoid conflicts between the MSVC debug and release runtime libraries, when
-  compiling a debug build of your application, you may need to link against a
-  debug build of libprotobuf.lib.  Similarly, release builds should link against
-  release libs.
+This directory contains *CMake* files that can be used to build protobuf
+with *MSVC* on *Windows*. You can build the project as a from *Command Prompt*
```
Comment (xfxyjwf) [inline cmake/README.md]: "You can build the project ~~as a~~ from *Command Prompt*"

File: cmake/README.md [hunk 2]
```diff
@@ -1,60 +1,243 @@
...
+and using an *Visual Studio* IDE.
+
+You need to have [CMake](http://www.cmake.org), [Visual Studio](https://www.visualstudio.com)
+and optionaly [Git](http://git-scm.com) installed on your computer before proceeding.
```
Comment (xfxyjwf) [inline cmake/README.md]: optionaly => optionally

File: cmake/README.md [hunk 3]
```diff
@@ -1,60 +1,243 @@
...
+
+Most of the instruction will be given to the *Сommand Prompt*, but the same
```
Comment (xfxyjwf) [inline cmake/README.md]: instruction => instructions

File: cmake/README.md [hunk 4]
```diff
@@ -1,60 +1,243 @@
...
+actions can be performed using appropriate GUI tools.
+
+Setting Environment
```
Comment (xfxyjwf) [inline cmake/README.md]: Environment Setup

File: cmake/README.md [hunk 5]
```diff
@@ -1,60 +1,243 @@
...
+===================
+
+From *Start* menu button open appropriate *Command Prompt*.
```
Comment (xfxyjwf) [inline cmake/README.md]: => "Open the appropriate *Command Prompt* from the *Start* menu."

File: cmake/README.md [hunk 6]
```diff
@@ -1,60 +1,243 @@
...
+
+For example: *Start* > *Visual Studio 2013* > *Visual Studio Tools* >
+*VS2013 x64 Native Tools Command Prompt*. You should see the *Command Prompt*:
+
+    C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\amd64>
+
+Change current drive letter, if needed:
```
Comment (xfxyjwf) [inline cmake/README.md]: These are too detailed steps that's better to be omitted. The readers are supposed to already know how to do that (i.e., change the working directory to a different driver).

Just say:
```
Change to your working directory:
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\amd64>cd C:\path\to\your\dir
C:\path\to\your\dir
```

File: cmake/README.md [hunk 7]
```diff
@@ -1,60 +1,243 @@
...
+    X:\>
+
+Where *X* is your working drive letter. And change directory to your working folder:
+
+    X:\>cd Path\to
+    X:\Path\to>
+
+Where *Path\to* is path to your real working folder.
+
+Make folder where builded protobuf will be installed:
```
Comment (xfxyjwf) [inline cmake/README.md]: Create a folder where protobuf headers/libraries/binaries will be installed after built.

File: cmake/README.md [hunk 8]
```diff
@@ -1,60 +1,243 @@
...
+
+    X:\Path\to>mkdir install
+
+If *cmake* coomand is not avaliable from *Command Promt*, add it in system *PATH* variable:
```
Comment (xfxyjwf) [inline cmake/README.md]: "add it in system ..." => "add it to system ..."

File: cmake/README.md [hunk 9]
```diff
@@ -1,60 +1,243 @@
...
+
+    X:\Path\to>set PATH=%PATH%;C:\Program Files (x86)\CMake\bin
+
+If *git* coomand is not avaliable from *Command Promt*, add it in system *PATH* variable:
```
Comment (xfxyjwf) [inline cmake/README.md]: "add it in ..." => "add it to .."

File: cmake/README.md [hunk 10]
```diff
@@ -1,60 +1,243 @@
...
+Good. Now you are ready to continue.
+
+Getting Sources
+===============
+
+You can get the latest stable source packages from the
+[releases](https://github.com/google/protobuf/releases) page.
+Or you can type:
+
+     X:\Path\to> git clone -b release_tag https://github.com/google/protobuf.git
```
Comment (xfxyjwf) [inline cmake/README.md]: Change release_tag to [release_tag] to indicate it's not meant to be exact text.

File: cmake/README.md [hunk 11]
```diff
@@ -1,60 +1,243 @@
...
+
+Where *release_tag* is a real repository tag or branch name as a *master*
```
Comment (xfxyjwf) [inline cmake/README.md]: => "Where *release_tag* is a git tag like v3.0.0-beta-1 or a branch name like *master*"

File: cmake/README.md [hunk 12]
```diff
@@ -1,60 +1,243 @@
...
+if you want to get the latest code.
+
+Go to the project folder:
+
+     X:\Path\to>cd protobuf
+     X:\Path\to\protobuf>
+
+Check whether a *gmock* directory exists in the upper level directory. If you
```
Comment (xfxyjwf) [inline cmake/README.md]:
```
Protobuf unit-tests require gmock to build. If you download protobuf source code from the releases page, the
gmock directory should already be there. If you checkout the code via "git clone", this gmock directory won't
exist and you will have to download it manually or skip building protobuf unit-tests.
```

File: cmake/README.md [hunk 13]
```diff
@@ -1,60 +1,243 @@
...
+
+     X:\Path\to\protobuf\gmock>cd ..\cmake
+     X:\Path\to\protobuf\cmake>
+
+Good. Now you are ready to *CMake* configuration.
+
+CMake Configuration
+===================
+
+The *CMake* provides a lot of different
```
Comment (xfxyjwf) [inline cmake/README.md]: "The CMake provides" => "CMake supports"

File: cmake/README.md [hunk 14]
```diff
@@ -1,60 +1,243 @@
...
+[generators](http://www.cmake.org/cmake/help/latest/manual/cmake-generators.7.html)
+for various native build system.
```
Comment (xfxyjwf) [inline cmake/README.md]: system => systems

File: cmake/README.md [hunk 15]
```diff
@@ -1,60 +1,243 @@
...
+We are interested a
```
Comment (xfxyjwf) [inline cmake/README.md]: "We are only interested in Makefile and Visual Studio generatrors."

File: cmake/README.md [hunk 16]
```diff
@@ -1,60 +1,243 @@
...
+[Makefile](http://www.cmake.org/cmake/help/latest/manual/cmake-generators.7.html#makefile-generators)
+and a
+[Visual Studio](http://www.cmake.org/cmake/help/latest/manual/cmake-generators.7.html#visual-studio-generators)
+generators.
+
+We will use shadow building to separate the temporary files from the protobuf source code.
+
+Make temporary *build* folder and change directory to:
```
Comment (xfxyjwf) [inline cmake/README.md]: "Create a temporary build folder and change your working directory to it:"

File: cmake/README.md [hunk 17]
```diff
@@ -1,60 +1,243 @@
...
+
+     X:\Path\to\protobuf\cmake>mkdir build & cd build
+     X:\Path\to\protobuf\cmake\build>
+
+The *Makefile* generator can build the project in only one configuration, so you need to build
```
Comment (xfxyjwf) [inline cmake/README.md]: Can we just describe how to build with Visual Studio IDE and remove the nmake stuff? The current description might confuse users and gives them a wrong impression that they need to follow all these steps.

File: cmake/README.md [hunk 18]
```diff
@@ -1,60 +1,243 @@
...
+To make *Visual Studio* solution file from *build* folder type:
+
+     X:\Path\to\protobuf\cmake>mkdir solution & cd solution
+     X:\Path\to\protobuf\cmake\solution>cmake -G "Visual Studio 12 2013 Win64" ^
+     -DCMAKE_INSTALL_PREFIX=../../../../install ^
+     ../..
+
+It will generate *Visual Studion* solution file *protobuf.sln* in current directory.
+
+If the *gmock* directory does not existence, and you do not want to build protobuf unit tests,
```
Comment (xfxyjwf) [inline cmake/README.md]: existence => exist

File: cmake/README.md (hunk 1 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 2 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 3 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 4 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 5 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed with some changes

File: cmake/README.md (hunk 6 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed with some changes

File: cmake/README.md (hunk 7 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 8 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 9 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 10 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 11 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 12 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 13 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 14 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 15 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 16 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 17 above)
Comment (podsvirov) [inline cmake/README.md]: Hmm .. Maybe it is for someone to be easier, but it then has to be
another instruction that does not use the command line, as it was laid down at the creation of this document.
Other instructions should only use GUI tools and probably contain screenshots.

File: cmake/README.md (hunk 18 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

**Comment by podsvirov** [comment]
@xfxyjwf I finished work on the notes and look forward to the next steps.
Please view updated documentation.

File: cmake/README.md [hunk 19]
```diff
@@ -1,60 +1,235 @@
...
+
+     C:\Path\to>cd protobuf
+     C:\Path\to\protobuf>
+
+Protobuf unit-tests require gmock to build. If you download protobuf source code
+from the *releases* page, the *gmock* directory should already be there. If you checkout
+the code via `git clone`, this *gmock* directory won't exist and you will have to
+download it manually or skip building protobuf unit-tests.
+
+If even so the *gmock* folder does not exist, you can download it as follows:
```
Comment (xfxyjwf) [inline cmake/README.md]: => "You can download gmock as follows:"

File: cmake/README.md [hunk 20]
```diff
@@ -1,60 +1,235 @@
...
+
+     C:\Path\to\protobuf\gmock>cd ..\cmake
+     C:\Path\to\protobuf\cmake>
+
+Good. Now you are ready to *CMake* configuration.
+
+CMake Configuration
+===================
+
+The *CMake* supports a lot of different
```
Comment (xfxyjwf) [inline cmake/README.md]: Remove the leading "The".

File: cmake/README.md (hunk 19 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 20 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md [hunk 21]
```diff
@@ -1,60 +1,235 @@
...
+
+     C:\Path\to\protobuf\cmake\build>mkdir release & cd release
+     C:\Path\to\protobuf\cmake\build\release>cmake -G "NMake Makefiles" ^
+     -DCMAKE_BUILD_TYPE=Release ^
+     -DCMAKE_INSTALL_PREFIX=../../../../install ^
+     ../..
+
+It will generate *nmake* *Makefile* in current directory.
+
+To using a *Debug* configuration from *build* folder type:
```
Comment (xfxyjwf) [inline cmake/README.md]: => To use "Debug” configuration:

File: cmake/README.md [hunk 22]
```diff
@@ -1,60 +1,235 @@
...
+
+     C:\Path\to\protobuf\cmake\build>mkdir debug & cd debug
+     C:\Path\to\protobuf\cmake\build\debug>cmake -G "NMake Makefiles" ^
+     -DCMAKE_BUILD_TYPE=Debug ^
+     -DCMAKE_INSTALL_PREFIX=../../../../install ^
+     ../..
+
+It will generate *nmake* *Makefile* in current directory.
+
+To make *Visual Studio* solution file from *build* folder type:
```
Comment (xfxyjwf) [inline cmake/README.md]: => To create Visual Studio solution file:

File: cmake/README.md [hunk 23]
```diff
@@ -1,60 +1,235 @@
...
+
+It will generate *Visual Studion* solution file *protobuf.sln* in current directory.
+
+If the *gmock* directory does not exist, and you do not want to build protobuf unit tests,
+you need to add *cmake* command argument `-Dprotobuf_BUILD_TESTS=OFF` to disable testing.
+
+Compiling
+=========
+
+To compilling protobuf type:
```
Comment (xfxyjwf) [inline cmake/README.md]: => To compile protobuf:

File: cmake/README.md [hunk 24]
```diff
@@ -1,60 +1,235 @@
...
+
+     C:\Path\to\protobuf\cmake\build\release>nmake
+
+or
+
+     C:\Path\to\protobuf\cmake\build\debug>nmake
+
+And wait for compiling to finish.
```
Comment (xfxyjwf) [inline cmake/README.md]: => And wait for the compilation to finish.

File: cmake/README.md [hunk 25]
```diff
@@ -1,60 +1,235 @@
...
+
+You prefer to use the IDE:
+
+  * Open the generated protobuf.sln file in Microsoft Visual Studio.
+  * Choose "Debug" or "Release" configuration as desired.
+  * From the Build menu, choose "Build Solution".
+
+And wait for compiling to finish.
```
Comment (xfxyjwf) [inline cmake/README.md]: wait for the compilation to finish.

File: cmake/README.md [hunk 26]
```diff
@@ -1,60 +1,235 @@
...
+
+Testing
+=======
+
+To start unit testing type:
```
Comment (xfxyjwf) [inline cmake/README.md]: => To run unit-tests:

File: cmake/README.md [hunk 27]
```diff
@@ -1,60 +1,235 @@
...
+
+     Running main() from gmock_main.cc
+     [==========] Running 1546 tests from 165 test cases.
+     
+     ...
+     
+     [==========] 1546 tests from 165 test cases ran. (2529 ms total)
+     [  PASSED  ] 1546 tests.
+
+Fore more specific testing start *tests* from source folder:
```
Comment (xfxyjwf) [inline cmake/README.md]: To run specific tests:

File: cmake/README.md [hunk 28]
```diff
@@ -1,60 +1,235 @@
...
+     [       OK ] AnyTest.TestPackAndUnpackAny (0 ms)
+     [ RUN      ] AnyTest.TestIs
+     [       OK ] AnyTest.TestIs (0 ms)
+     [----------] 3 tests from AnyTest (1 ms total)
+     
+     [----------] Global test environment tear-down
+     [==========] 3 tests from 1 test case ran. (2 ms total)
+     [  PASSED  ] 3 tests.
+
+Note that the start from source folder important for testing.
```
Comment (xfxyjwf) [inline cmake/README.md]: => Note that the tests must be run from the source folder.

File: cmake/README.md [hunk 29]
```diff
@@ -1,60 +1,235 @@
...
+
+If all tests are passed, safely continue.
+
+Installing
+==========
+
+To install protobuf to the specified *install* folder type:
```
Comment (xfxyjwf) [inline cmake/README.md]: folder type => folder (remove type)

File: cmake/README.md [hunk 30]
```diff
@@ -1,60 +1,235 @@
...
+     C:\Path\to\protobuf\cmake\build\release>nmake install
+
+or
+
+     C:\Path\to\protobuf\cmake\build\debug>nmake  install
+
+You can also build project *INSTALL* from Visual Studio solution.
+It sounds not so strange and it works.
+
+This will create the following folders:
```
Comment (xfxyjwf) [inline cmake/README.md]: This will create the following folders under the install location:

File: cmake/README.md [hunk 31]
```diff
@@ -1,60 +1,235 @@
...
+  * bin - that contain protobuf *protoc.exe* compiler;
```
Comment (xfxyjwf) [inline cmake/README.md]: contain => contains

Same for the following two lines.

File: cmake/README.md [hunk 32]
```diff
@@ -1,60 +1,235 @@
...
+  * inclue - that contain C++ headers and protobuf *.proto files;
+  * lib - that contain linking libraries and *CMake* configuration for *protobuf* package.
```
Comment (xfxyjwf) [inline cmake/README.md]: CMake configuration => CMake configuration files

File: cmake/README.md [hunk 33]
```diff
@@ -1,60 +1,235 @@
...
+
+Now you can if needed:
+  * Copy the contents of the include directory to wherever you want to put headers.
+  * Copy protoc.exe wherever you put build tools (probably somewhere in your PATH).
+  * Copy linking libraries libprotobuf[d].lib, libprotobuf-lite[d].lib, and libprotoc[d].lib wherever you put libraries.
```
Comment (xfxyjwf) [inline cmake/README.md]: Do we have the "d" suffix on debug-built libraries? I don't remember we added any.

File: cmake/README.md [hunk 34]
```diff
@@ -90,21 +262,43 @@ ZLib support
...
 
-        $ cmake -G "Visual Studio 9 2008" -DZLIB=ON ..
+  * Make sure zlib's two headers are in your `C:\Path\to\install\include` path
+  * Make sure zlib's linking libraries (*.lib file) is in your
+    `C:\Path\to\install\lib` library path.
 
-   If it reports NOTFOUND for zlib_include or zlib_lib, you might haven't put
-   the headers or the .lib file in the right directory.
-4) Open the generated protobuf.sln file and build as usual.
+You can also quick compile it himself.
```
Comment (xfxyjwf) [inline cmake/README.md]: => You can also compile it from source by yourself.

**Comment by xfxyjwf** [comment]
Done with my comments:) Let me know when you finish updating this PR.

File: cmake/README.md (hunk 33 above)
Comment (podsvirov) [inline cmake/README.md]: In #783 I added the option protobuf_DEBUG_POSTFIX, which by default is "d" and can be changed if necessary.

File: cmake/README.md (hunk 21 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 22 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 23 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 24 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 25 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 26 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 27 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 28 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 29 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 30 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 31 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 32 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md (hunk 34 above)
Comment (podsvirov) [inline cmake/README.md]: Fixed

File: cmake/README.md [hunk 35]
```diff
@@ -90,21 +262,46 @@ ZLib support
...
+protobuf project.
+
+Now add *bin* folder from *install* to system *PATH*:
+
+     C:\Path\to>set PATH=%PATH%;C:\Path\to\install\bin
+
+You need reconfigure protobuf with flag `-Dprotobuf_WITH_ZLIB=ON` when invoking cmake.
+
+Note that if you have compiled ZLIB youself, as stated above,
+further disable the option `-Dprotobuf_MSVC_STATIC_RUNTIME=OFF`.
```
Comment (podsvirov) [inline cmake/README.md]: @xfxyjwf please, review this new paragraph.

**Comment by podsvirov** [comment]
@xfxyjwf I finished work on the notes and look forward to the next steps.
Please view updated documentation.

File: cmake/README.md (hunk 35 above)
Comment (xfxyjwf) [inline cmake/README.md]: youself => yourself

**Comment by xfxyjwf** [comment]
@podsvirov Could you squash the commits into a single one? I'll then merge the PR to master.

**Comment by podsvirov** [comment]
@xfxyjwf Good. Now it is one commit.

**Comment by podsvirov** [comment]
@xfxyjwf Excellent! But to be honest, I thought the first #783 PR will be merged?...

**Pull request merged**